sim/core/vengeance.go

```diff
@@ -0,0 +1,104 @@
+package core
+
+import (
+	"math"
+	"time"
+
+	"github.com/wowsims/cata/sim/core/stats"
+)
+
+type VengeanceTracker struct {
+	eligibleDamage   float64
+	apBonus          float64
+	prevAPBonus      float64
+	recentMaxAPBonus float64
+	lastAttackedTime time.Duration // timestamp that the character was last attacked
+}
+
+const (
+	VengeanceAPDecayRate     = 0.1 // AP bonus decays by 10% every 2 seconds, or 5% if the character has been hit in that time
+	OutcomeVengeanceTriggers = OutcomeLanded
+)
+
+func Clamp(val float64, min float64, max float64) float64 {
+	return math.Max(min, math.Min(val, max))
+}
+
+func UpdateVengeance(sim *Simulation, character *Character, tracker *VengeanceTracker, aura *Aura) {
+	// Save the current AP bonus so we can apply the new buff correctly
+	tracker.prevAPBonus = tracker.apBonus
+
+	// If this character has been attacked in the last 2 seconds, apply half decay and add new damage to buff
+	timeSinceLastHit := sim.CurrentTime - tracker.lastAttackedTime
+	if timeSinceLastHit < time.Second*2 {
+
+		// Decay existing bonus by half of the rate
+		decay := VengeanceAPDecayRate / 2
+		tracker.apBonus -= (decay * tracker.apBonus)
+
+		// Add 5% of damage taken in the last 2 seconds
+		tracker.apBonus += 0.05 * tracker.eligibleDamage
+
+		// 4.3.0 change: the vengeance AP buff is always at least 33% of the incoming
+		// damage if the tank has been hit in the last 2 seconds
+		baseAPBonus := tracker.eligibleDamage / 3.0
+		tracker.apBonus = math.Max(tracker.apBonus, baseAPBonus)
+	} else {
+		// No hits in the last 2 seconds - apply full decay
+		tracker.apBonus -= (VengeanceAPDecayRate * tracker.recentMaxAPBonus)
+	}
+
+	// TODO: verify that stam actually shows up in this formula, currently implemented as simcraft has it
+	apBonusMax := character.GetStat(stats.Stamina) + 0.1*character.GetStat(stats.Health)
```

Comment:
I think we prefer if we do any health queries trough the HealthBar so change `character.GetStat(stats.Health)` to `character.MaxHealth()`

Reply:
Oh duh, that's what I get for writing code at 6am. Fixedd!